### PR TITLE
Jest tests for search page pagination and sorting

### PR DESCRIPTION
**Pull request by PenghaiZhang**
#1306 

##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [ ] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

Update search page Jest test, including: 
1. adding new tests for Search page pagination and the sorting control; and
2. refactoring the tests for spinner and debounce query; and
3. updating mock data.

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 1]
```diff
@@ -62,50 +66,97 @@ describe("<SearchPage/>", () => {
...
-    changeQuery("old title");
-    //use a timed callback to wait for the debounce before asserting results have populated the page
-    setTimeout(() => {
-      expect(component.html()).toContain("No results found.");
-    }, 1000);
+  it("should support debounce query search", async () => {
+    jest.useFakeTimers("modern");
+    const input = component.find("input.MuiInputBase-input");
+    await updateSearchOptions(() =>
+      input.simulate("change", { target: { value: "new query" } })
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:94]: Help wanted here. 
Warning `An update to SearchPage inside a test was not wrapped in act(...).` is always given. 
I can see it is caused by debouncing search query which triggers a state update after 500ms.  I have not worked out how to put this update inside a `act`.

Comment (ChristianMurphy) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:94]: Would moving `jest.advanceTimersByTime` inside the `act` resolve the issue :thinking:

Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:94]: Thanks @ChristianMurphy ! It does resolve the issue!

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 2]
```diff
@@ -62,50 +66,94 @@ describe("<SearchPage/>", () => {
...
-  it("should display 'No results found.' when there are no search results", () => {
-    changeQuery("old title");
-    //use a timed callback to wait for the debounce before asserting results have populated the page
-    setTimeout(() => {
-      expect(component.html()).toContain("No results found.");
-    }, 1000);
+  it("should support debounce query search", async () => {
+    jest.useFakeTimers("modern");
+    const input = component.find("input.MuiInputBase-input");
+    await updateSearchOptions(() => {
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:90]: do these await simulate blocks need to be surrounded with an `act` function? Whenever I did things like that I'd get warnings in the jest console.

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 3]
```diff
@@ -62,50 +66,94 @@ describe("<SearchPage/>", () => {
...
+  });
+
+  it("should support sorting search results", async () => {
+    const sortingControl = component.find(".MuiCardHeader-action input");
+    await updateSearchOptions(() =>
+      sortingControl.simulate("change", {
+        target: { value: SortOrder.DATEMODIFIED },
+      })
+    );
+    // Because sorting is done on Server and we are using mock data, we can only check if the selected
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:138]:
```suggestion
    // Because sorting is done on the server-side and we are using mock data, we can only check if the selected
```

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 4]
```diff
@@ -62,50 +66,94 @@ describe("<SearchPage/>", () => {
...
+    });
   });
 
   it("should display a spinner when search is in progress", async () => {
-    // Trigger a search by changing search query.
-    changeQuery("new query");
-    setTimeout(async () => {
-      expect(component.find(CircularProgress)).toHaveLength(1);
-    }, 1000);
+    // Trigger a search by changing sorting order.
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:147]:
```suggestion
    // Trigger a search by changing sorting order
```

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 5]
```diff
@@ -62,50 +66,94 @@ describe("<SearchPage/>", () => {
...
+
+  it("should support sorting search results", async () => {
+    const sortingControl = component.find(".MuiCardHeader-action input");
+    await updateSearchOptions(() =>
+      sortingControl.simulate("change", {
+        target: { value: SortOrder.DATEMODIFIED },
+      })
+    );
+    // Because sorting is done on Server and we are using mock data, we can only check if the selected
+    // sort order is included in the search params.
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:139]:
```suggestion
    // sort order is included in the search params
```

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 6]
```diff
@@ -62,50 +66,94 @@ describe("<SearchPage/>", () => {
...
-  it("should contain the test data after a search bar text change and render", () => {
-    changeQuery("new title");
-    setTimeout(() => {
-      expect(component.html()).not.toContain("No results found.");
-      expect(component.html()).toContain(
-        "266bb0ff-a730-4658-aec0-c68bbefc227c"
-      );
-    }, 1000);
+  it("should support changing the number of items displayed per page", async () => {
+    // Initial items per page is 10.
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:103]:
```suggestion
    // Initial items per page is 10
```

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 7]
```diff
@@ -62,50 +66,94 @@ describe("<SearchPage/>", () => {
...
-      expect(component.html()).toContain("No results found.");
-    }, 1000);
+  it("should support debounce query search", async () => {
+    jest.useFakeTimers("modern");
+    const input = component.find("input.MuiInputBase-input");
+    await updateSearchOptions(() => {
+      input.simulate("change", { target: { value: "new query" } });
+      jest.advanceTimersByTime(1000);
+    });
+    // After 1s the second search should be triggered.
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:94]:
```suggestion
    // After 1s the second search should be triggered
```

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx (hunk 2 above)
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:90]: Yes they need. `updateSearchOptions` takes a function as its argument and executes the function inside an `act`.

Comment (SammyIsConfused) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:90]: Oh I see now, its in the function rather than up here. 
Cool! Looks good to me.

**Comment by SammyIsConfused** [comment]
Thanks for sorting those out! Good to go.

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 8]
```diff
@@ -62,50 +66,94 @@ describe("<SearchPage/>", () => {
...
-      const input = component.find("input.MuiInputBase-input");
-      input.simulate("change", { target: { value: query } });
-    });
-  };
+  /**
+   * Update search options in order to trigger a search.
+   * @param update A function that simulates UI behaviours such as selecting a different value from a dropdown.
+   */
+  const updateSearchOptions = async (update: () => void) =>
+    await act(async () => await update());
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:77]: isn't
wouldn't calling
```ts
updateSearchOptions(() => itemsPerPageSelect.simulate("change", { target: { value: 25 } }));
```

be the same as

```ts
act(() => itemsPerPageSelect.simulate("change", { target: { value: 25 } }));
```

?

I'm missing what this wrapper around `act` is adding.

Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:77]: With `act(() => itemsPerPageSelect.simulate("change", { target: { value: 25 } }));` , the warning `An update to SearchPage inside a test was not wrapped in act(...).` is always given.  So an async act can resolve this warning. 

    await act(async () => {
      await itemsPerPageSelect.simulate("change", { target: { value: 25 } })
    });

Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:77]: So is this just to reduce verbosity? If so I do wonder if it should be called `awaitAct` or some-such, as it would appear to have no direct knowledge of anything to do with search.

But also, what's the net gain here? The removal of one use of the `async` and `await` keywords? If so, maybe it's not needed, but let me throw in an example for our consideration:

```typescript
    await updateSearchOptions(() => {
      input.simulate("change", { target: { value: "new query" } });
      jest.advanceTimersByTime(1000);
    });
```

Is this simply the same as:

```typescript
    await act( async () => {
      input.simulate("change", { target: { value: "new query" } });
      await jest.advanceTimersByTime(1000);
    });
```

?

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 9]
```diff
@@ -62,50 +66,94 @@ describe("<SearchPage/>", () => {
...
-  it("should retrieve search settings and do a search when the page is opened", async () => {
+  it("should retrieve search settings and do a search when the page is opened", () => {
     expect(
       SearchSettingsModule.getSearchSettingsFromServer
     ).toHaveBeenCalledTimes(1);
     expect(SearchModule.searchItems).toHaveBeenCalledTimes(1);
+    expect(SearchModule.searchItems).toHaveBeenCalledWith(defaultSearchOptions);
   });
 
-  it("should display 'No results found.' when there are no search results", () => {
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:82]: I don't seen an equivalent test for this? Why is it removed?

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 10]
```diff
@@ -62,50 +66,94 @@ describe("<SearchPage/>", () => {
...
+    });
+    // After 1s the second search should be triggered
+    expect(SearchModule.searchItems).toHaveBeenCalledTimes(2);
+    expect(SearchModule.searchItems).toHaveBeenCalledWith({
+      ...defaultSearchOptions,
+      query: "new query",
+    });
   });
 
-  it("should contain the test data after a search bar text change and render", () => {
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:90]: I don't see an equivalent test for this (especially one that actually verifies that search results are in the list). Why the removal?

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx (hunk 9 above)
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:82]: Because we are using mock data, the result will always be the mock data no matter  whatever search query we use.

Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:82]: expect(SearchModule.searchItems).toHaveBeenCalledWith({
      ...defaultSearchOptions,
      query: "new query",
    });

This expect can make sure that the search query has been updated.

Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:82]: > Because we are using mock data, the result will always be the mock data no matter whatever search query we use.

So then, should we have a case where we mock with empty data to test the empty data result? It's arguably an important case to test - being a boundary condition.

> This expect can make sure that the search query has been updated.

Not sure which question that's in response to. :thinking:

Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:82]: Yes I will add one more test for empty data result!

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 11]
```diff
@@ -97,6 +100,21 @@ describe("<SearchPage/>", () => {
...
+  it("should display 'No results found.' when there are no search results", async () => {
+    mockSearch.mockImplementationOnce(() =>
+      Promise.resolve(getEmptySearchResult)
+    );
+    jest.useFakeTimers("modern");
+    const input = component.find("input.MuiInputBase-input");
+    await awaitAct(() => {
+      input.simulate("change", { target: { value: "no items" } });
+      jest.advanceTimersByTime(1000);
+    });
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:116]: This is now duplicated. Seems you need a function of:

```typescript
const doSearch = async (searchTerm: string) => {
    jest.useFakeTimers("modern");
    const input = component.find("input.MuiInputBase-input");
    await awaitAct(() => {
      input.simulate("change", { target: { value: searchTerm } });
      jest.advanceTimersByTime(1000);
    });
}
```
